### PR TITLE
Better support for node restarts

### PR DESCRIPTION
**Pull request by mystenmark**
- Ensure that all tasks owned by a killed node are immediately destroyed so that all drop handlers run promptly.
- Provide the `return_if_killed!()` macro to allow drop handlers to return early to better simulate a crash.

### CODE REVIEW

File: msim/src/sim/net/mod.rs [hunk 1]
```diff
@@ -177,6 +182,23 @@ impl HostNetworkState {
...
         Ok(cb(socket))
     }
+
+    fn delete_node(&mut self, id: NodeId) {
+        let to_remove: Vec<_> = self
+            .sockets
+            .iter()
+            .filter_map(|((node, fd), _)| {
+                if *node == id {
+                    Some((*node, *fd))
```
Comment (williampsmith) [inline msim/src/sim/net/mod.rs:192]: style nit - feel free to ignore: why not just add these lines in the for loop and remove the `filter_map`?

File: msim/src/sim/net/network.rs [hunk 2]
```diff
@@ -108,6 +109,32 @@ impl Network {
...
+            self.addr_to_node.remove(ip);
+        }
+        self.clogged_node.remove(&id);
+
+        let to_remove: Vec<_> = self
+            .clogged_link
+            .iter()
+            .filter_map(|(a, b)| {
+                if *a == id || *b == id {
+                    Some((*a, *b))
```
Comment (williampsmith) [inline msim/src/sim/net/network.rs:126]: Same nit as above

File: msim/src/sim/runtime/mod.rs [hunk 3]
```diff
@@ -508,12 +517,40 @@ impl Future for NodeHandle {
...
+        if task.is_killed() {
+            trace!("current task is killed");
+            return true;
+        }
+    } else {
+        trace!("no current task");
+        return true;
+    }
+    return false;
+}
```
Comment (williampsmith) [inline msim/src/sim/runtime/mod.rs]: nit - take it or leave it: This seems more readable (not having to reason about the fallthrough case) and more idiomatic to me
```suggestion
        if task.is_killed() {
            trace!("current task is killed");
            true
        } else {
            false
        }
    } else {
        trace!("no current task");
        true
    }
}
```

**Review by williampsmith — APPROVED** [review]

File: msim/src/sim/net/mod.rs (hunk 1 above)
Comment (mystenmark) [reply msim/src/sim/net/mod.rs:192]: because there's no easy way to modify a container while iterating over it

Comment (williampsmith) [reply msim/src/sim/net/mod.rs:192]: oh yes, my mistake :)

File: msim/src/sim/net/network.rs (hunk 2 above)
Comment (mystenmark) [reply msim/src/sim/net/network.rs:126]: same reply 😄

File: msim/src/sim/runtime/mod.rs (hunk 3 above)
Comment (mystenmark) [reply msim/src/sim/runtime/mod.rs]: fixed

**Pull request merged**